src/libasr/codegen/wasm_to_x86.cpp

```diff
@@ -0,0 +1,434 @@
+#include <fstream>
+#include <chrono>
+#include <iomanip>
+
+#include <libasr/assert.h>
+#include <libasr/codegen/wasm_to_x86.h>
+#include <libasr/codegen/x86_assembler.h>
+
+// #define WAT_DEBUG
+
+#ifdef WAT_DEBUG
+#define DEBUG(s) std::cout << s << std::endl
+#else
+#define DEBUG(s)
+#endif
+
+namespace LFortran {
+
+namespace {
+
+// This exception is used to abort the visitor pattern when an error occurs.
+class CodeGenAbort {};
+
+// Local exception that is only used in this file to exit the visitor
+// pattern and caught later (not propagated outside)
+class CodeGenError {
+   public:
+    diag::Diagnostic d;
+
+   public:
+    CodeGenError(const std::string &msg)
+        : d{diag::Diagnostic(msg, diag::Level::Error, diag::Stage::CodeGen)} {}
+
+    CodeGenError(const std::string &msg, const Location &loc)
+        : d{diag::Diagnostic(msg, diag::Level::Error, diag::Stage::CodeGen,
+                             {diag::Label("", {loc})})} {}
+};
+
+}  // namespace
+
+namespace wasm {
+
+class WASMDecoder {
```

Comment:
This `WASMDecoder` class is being repeated (it is already defined in `src/libasr/codegen/wasm_to_wat.cpp`). We should extract it and place in a common file.

**doubt:** Where should we place this `WASMDecoder` class? Creating a new file (e.g. `wasm_disassembler.cpp` or `wasm_decoder.cpp`) seems possible, but I am unsure if this is the right way.

Reply:
I temporarily moved this class into a new file `wasm_decoder.h`.